### ISSUE TITLE
psycopg2.ProgrammingError exception did not return all the errors when creating procedure with multiple errors on EDB Postgres Advanced Server

### ISSUE THREAD

**Issue by akshay-joshi** [psycopg/psycopg2#855]
Hello

I am using psycopg2 2.7.7 to connect to the EDB Postgres Advanced Server using asynchronous support (async=1). When I ran the following procedure:
CREATE OR REPLACE PROCEDURE compile_test
AS
v_day DATE;
v_next_day DATE;
BEGIN
SELECT NOW ITO v_day;
SELECT NOW()+1 Day INTO v_next_day;
DBMS_OUTPUT.PUT_LINE(v_day);
DBMS_OUTPUT.PUT_LINE(v_next_day);
END;

**on PSQL Prompt**: It is throwing all the errors when run on psql prompt. Refer the screenshot

<img width="940" alt="psql_prompt" src="https://user-images.githubusercontent.com/1632882/53715943-62bc8e00-3e79-11e9-939e-8376ae55fb71.png">
**on Python Prompt using psycpog2**: It returns only one error "ERROR: compilation of SPL function/procedure "compile_test" failed due to 2 errors SQL state: 42601". Which is the last error of the psql prompt. Refer the screenshot:

![psycopg2_multiple_error](https://user-images.githubusercontent.com/1632882/53715979-7bc53f00-3e79-11e9-8ef6-c9ddf0c97fd2.png)

How can we read all the errors when psycopg2.ProgrammingError exception occurs ? Can anyone please provide some pointers, suggestions or it is a bug in psycopg2?

**Comment by dvarrazzo** [comment]
I have no idea. psycopg doesn't support EDB officially. There must be some difference they have in the line protocol we are not aware of.

EDB is a commercial product. You will have to speak with them to have support for the requested feature.

**Comment by akshay-joshi** [comment]
Hi Daniele

I think it's nothing to do with EDB's commercial product. When I ran the same query in Synchronous mode it returns all the errors, but in case of async=1 it returns only the last error.

To test this I have compiled psycopg2 in DEBUG mode and first test the same query in synchronous mode. Refer the below screenshot
<img width="700" alt="sync_connection" src="https://user-images.githubusercontent.com/1632882/53885639-428cfa80-4044-11e9-9a65-d0d028183004.png">

When I run it with async=1, it returns only the last error. Refer the below screenshot:
<img width="1021" alt="async_connection" src="https://user-images.githubusercontent.com/1632882/53885695-62bcb980-4044-11e9-9b5a-c3f70121b17c.png">

I have check the source code and found one comment in **pq_get_last_result** function of **pqpath.c** where it is mentioned discarding the result from all the queries except the last. So could it be the possible reason?? Refer the below screenshot
<img width="629" alt="pqpath c" src="https://user-images.githubusercontent.com/1632882/53886042-23429d00-4045-11e9-968c-02c5cc481b0f.png">

**Comment by dvarrazzo** [comment]
Uhm, looks like you are right. Incidentally, I'm just refactoring that function because of #856. I thought we were discarding the result but at least checking the error condition.

Amazing how there have been two issues found in a function untouched for so many years in the space of one day...

**Comment by dvarrazzo** [comment]
Still it seems to me your server behaves differently from postgres.

I tried creating a table:

```
create table foo(id integer primary key check (id > 0));
```

sending a sequence of operation, the server stops working at the first error:

```
In [27]: cur.execute("insert into foo values (1); insert into foo values (-1); insert into foo values (2)")
[13401] psyco_curs_execute: starting execution of new query
[13401] pq_execute: pg connection at 0x558dd4fe3160 OK
[13401] pq_begin_locked: pgconn = 0x558dd4fe3160, autocommit = 1, status = 1
[13401] pq_begin_locked: transaction in progress
[13401] pq_execute: executing ASYNC query: pgconn = 0x558dd4fe3160
[13401]     insert into foo values (1); insert into foo values (-1); insert into foo values (2)
[13401] pq_execute: async query sent to backend
[13401] pq_execute: query got flushed immediately
[13401] psyco_curs_execute: res = 0, pgres = (nil)

In [28]: wait_select(cnn)
[13401] conn_poll: status = 1
[13401] conn_poll: async_status = ASYNC_READ
[13401] pq_is_busy: consuming input
[13401] conn_poll: poll reading
[13401] conn_poll: async_status -> ASYNC_DONE
[13401] pq_get_last_result: got result PGRES_COMMAND_OK
[13401] pq_get_last_result: got result PGRES_FATAL_ERROR
[13401] pq_fetch: pgstatus = PGRES_FATAL_ERROR
[13401] pq_fetch: uh-oh, something FAILED: status = 7 pgconn = 0x7fb45730dc80
[13401] pq_raise: PQresultErrorMessage: err=ERROR:  new row for relation "foo" violates check constraint "foo_id_check"
DETAIL:  Failing row contains (-1).

[13401] pq_raise: err2=new row for relation "foo" violates check constraint "foo_id_check"
DETAIL:  Failing row contains (-1).

[13401] pq_fetch: fetching done; check for critical errors
```

You can see a `PGRES_COMMAND_OK`, a `PGRES_FATAL_ERROR`, and no further operations.

Does it happen the same with your server?

**Comment by akshay-joshi** [comment]
Hi Daniele

Yes the behaviour of the EDB Postgres Plus Advanced server is different then postgres. Database server have one configuration option which instructs the server to stop parsing if it encounters the specified number of errors in functions/procedures code, or when it encounters an error in SQL code. Default value of that parameter is 10, so it returns maximum 10 errors.

When I connect to the EDB's database server and try to create a function or procedure with multiple error statements, in **_Synchronous_** mode it works properly and return all the errors. For example
create table foo(id integer primary key check (id > 0));
CREATE OR REPLACE PROCEDURE public.tp()
AS
BEGIN
  insert ito foo1 values (-5);     // Error Statement
  insert itow foo2 values (-1);  // Error Statement
  insert into foo values (2);
END;

When I try to run the above in **_Asynchronous_** mode then it returns only the last error with **psycopg2**

I have tested it with **libpq** and in both mode it returns all the errors. Attached is the C file that I have used to test it with libpq. Please change the extension to .c from .txt and change the connection parameter as per your requirement.
[async_multiple_error.txt](https://github.com/psycopg/psycopg2/files/2944984/async_multiple_error.txt)

So even though behaviour of the database server is different, it behaves correctly in **_Synchronous_** mode and behaviour with **libpq** is also same. Seems like there is some problem in psycopg2 with **_Asynchronous_** call.

**Comment by dvarrazzo** [comment]
Hi @akshay-joshi 

even if we receive all the errors from the libpq, what can we do in python? We cannot raise two exceptions, so we have to choose what to raise.

The postgres behaviour, if you test for "error, error, ok", is to stop at the first error. So in all the tests I can perform there will always be only one error, and it will be on the last result returned. This is my universe. I understand you have a different server which returns two errors: your server is untested and unsupported, if its behaviour is different from the normal Postgres. I can do something on a best-effort base but cannot guarantee a thing. So far so good?

Now, even if I receive a sequence of results, two of whose are errros, what can a reasonable Python program should do? Upon the sequence:

- error1
- error2
- success

we currently, in the behaviour you observer, return a success.

- If we raise the exception of error1 we swallow error2 and we ignore the success.
- If we raise the exception of error2 we swallow error1 and we ignore the success.

Crucially, the behaviour of the program in sync and async modes will be different: in sync mode I understand your server will at error1. In async mode it will process all the statements, returning error results but continuing, and whether we raise error1 or error2 the success statement will be executed anyway, which would be very surprising.

Should we ask the server to stop processing at error1? there is a race condition: even if we sent a cancel message on the first error the database would have gone ahead anyway, and that would be a blocking operation I wouldn't do in async mode unless in case of an emergency (and an error is not an emergency, it's part of the normal process).

Raising at either error1 or error2 also cause the problem that there are following results on the pipe, so most likely the server and the client will go out of sync in their async communication: the function consuming the results should process all the results to the end before raising anything or the connection will likely end up in an unusable state.

So all in all:

- the behaviour of your database is different from postgres. Your database behaviour is not supported by psycopg.
- there is no way in python to raise two exceptions.
- there is no way to stop the server timely after the first exception.
- not processing all the results will likely break the connection.
- whether we report the first or the last exception, the database will have also processed operations we don't know about.

So, what would you suggest to do?

> Yes the behaviour of the EDB Postgres Plus Advanced server is different then postgres. Database server have one configuration option which instructs the server to stop parsing if it encounters the specified number of errors in functions/procedures code, or when it encounters an error in SQL code. Default value of that parameter is 10, so it returns maximum 10 errors.

psycopg is not designed to handle this. It is unclear how it can even support this if we convert errors into exceptions. I don't have a test EDB server and I have never received a request to support it.

Also note that I'm just in the process of changing the async communication in order to fix #856, which could result in the behaviour you are observing to change after the release of psycopg 2.8 (which could be stopping at the first error and breaking the connection).

**Comment by dvarrazzo** [comment]
In the above changeset there is now an algorithm which supposedly consumes the whole stream of async responses but raises an exception for the first. Please note that, for the reasons explained here, the behaviour on EDB Postgres Advanced Server is **untested**. Feel free to test and, if you have any feedback, I'll try to incorporate it.
